process/transaction/transactionCostEstimator.go

```diff
@@ -77,53 +92,76 @@ func (tce *transactionCostEstimator) simulateTransactionCost(tx *transaction.Tra
 	res, err := tce.txSimulator.ProcessTx(tx)
 	if err != nil {
 		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: err.Error(),
+			GasUnits:      0,
+			ReturnMessage: err.Error(),
 		}, nil
 	}
 
 	isMoveBalanceOk := txType == process.MoveBalance && res.FailReason == ""
 	if isMoveBalanceOk {
 		return &transaction.CostResponse{
-			GasUnits:   tce.feeHandler.ComputeGasLimit(tx),
-			RetMessage: "",
+			GasUnits:      tce.feeHandler.ComputeGasLimit(tx),
+			ReturnMessage: "",
 		}, nil
 
 	}
 
 	if res.FailReason != "" {
 		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: res.FailReason,
+			GasUnits:      0,
+			ReturnMessage: res.FailReason,
 		}, nil
 	}
 
 	if res.VMOutput == nil {
 		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: process.ErrNilVMOutput.Error(),
-			Scrs:       nil,
+			GasUnits:             0,
+			ReturnMessage:        process.ErrNilVMOutput.Error(),
+			SmartContractResults: nil,
 		}, nil
 	}
 
 	if res.VMOutput.ReturnCode == vmcommon.Ok {
 		return &transaction.CostResponse{
-			GasUnits:   tx.GasLimit - res.VMOutput.GasRemaining,
-			RetMessage: "",
-			Scrs:       res.ScResults,
+			GasUnits:             computeGasUnitsBasedOnVMOutput(tx, res.VMOutput),
+			ReturnMessage:        "",
+			SmartContractResults: res.ScResults,
 		}, nil
 	}
 
 	return &transaction.CostResponse{
-		GasUnits:   0,
-		RetMessage: fmt.Sprintf("%s %s", res.VMOutput.ReturnCode.String(), res.VMOutput.ReturnMessage),
-		Scrs:       res.ScResults,
+		GasUnits:             0,
+		ReturnMessage:        fmt.Sprintf("%s %s", res.VMOutput.ReturnCode.String(), res.VMOutput.ReturnMessage),
+		SmartContractResults: res.ScResults,
 	}, nil
 }
 
+func computeGasUnitsBasedOnVMOutput(tx *transaction.Transaction, vmOutput *vmcommon.VMOutput) uint64 {
+	isTooMuchGasProvided := strings.Contains(vmOutput.ReturnMessage, smartContract.TooMuchGasProvidedMessage)
+	if !isTooMuchGasProvided {
+		return tx.GasLimit - vmOutput.GasRemaining
+	}
+
+	return tx.GasLimit - extractGasRemainedFromMessage(vmOutput.ReturnMessage)
+}
+
+func extractGasRemainedFromMessage(message string) uint64 {
+	splitMessage := strings.Split(message, "gas remained = ")
+	if len(splitMessage) < 2 {
+		return 0
```

Comment:
this is an error case, maybe log the entire message as to easily debug? Also for L156. I would put these on WARN level because it is a programming error.

Reply:
+1

Reply:
done.